Allow UIElements to specify mouse cursor state when in dragged or mouse-over state.

abaga129 opened the pull request:
This isnt ready for merge yet.  I wanted to go ahead and open this PR to discuss it.

`UIElement` now has two properties `CursorWhenDragged` and `CursorWhenMouseOver`.  

These are retrieved by the UIContext using the `getCurrentMouseCursor` method.  This method prioritizes the element that is being dragged.  If no element is being dragged, it looks for the element being moused over, lastly if it will default to the `MouseCursor.pointer`.

The  MouseCursor enum can easily be changed to use different names, or to add/remove possible cursor types we might want.  Any thoughts on this?
```D
enum MouseCursor
{
    pointer, //default cursor
    linkSelect,
    move,
    drag,
    verticalResize,
    horizontalResize,
    hidden
}
```

Currently `X11Window` is working.  I still have to add the changes necessary for win32 and cocoa.

p0nce commented:
A few thoughts:
- I like the names of the enum and enum choices
- I like the polling on each redraw to see if the cursor changes. Avoids push notification for something that is easily cached on the windowing side. Great!
- Please do not use `@property`
- Please use camelCase for the methods like `CursorWhenDragged()` => `cursorWhenDragged()`
- if `_lastX11CursorFont` is initialized to zero intentionally https://github.com/AuburnSounds/Dplug/pull/482/commits/d2303fa954255c0425f079c42ead398684ef9782#diff-e37d7dbd655eb54ad077c52bfe378edfR265 maybe write ` = 0;`. Could also be `= 2;` . since the first cursor set will just set it again to "default". No idea.

This is almost perfect!

abaga129 commented:
Thanks for looking this over! I'm working on the suggested changes now.  All that should be left is the Win32 and cocoa implementations.

One change in going to make is to have the Window pass the x and y coordinates of the mouse to the listener when retrieving the cursor.  These will be passed down to the UIElements so we can have more complex behavior if we wish.

p0nce commented:
Hello getting back to you tomorrow!

p0nce commented:
```d
    /// Returns the cursor type that should be shown when this element
    /// is being dragged.
    /// This should be overriden to change default behavior
    /// `x` and `y` are the coordinates of the mouse when this is called
    MouseCursor cursorWhenDragged(int x, int y) 

    /// Returns the cursor type that should be shown when this element
    /// is being moused over.
    /// This should be overriden to change default behavior
    /// `x` and `y` are the coordinates of the mouse when this is called
    MouseCursor cursorWhenMouseOver(int x, int y)
```

I don't think it's legitimate that these function are called with the mouse position.

Widgets that want to display a selector for the subpart that would be selected if clicked often have internal state that know which subpart is "under mouse". That way you can draw a visual help. This require to set dirtyness, so there is a need for such a state already.

For `cursorWhenDragged`, I'm sure it's even less legitimate since the drag will carry on beyong the widgets boundaries (in which case your assert will yield false). This position doesn't identify the subpart that is dragged.

Work-around:

1) Maintain the latest mouse position
You can always do otherwise: onMouseMove, onMouseClick and onMouseDrag to update the current Mouse position (granted, this should be available with a UIElement final protected method to get the latest mouse position as stored in the context, but in local coordinates)

2) Maintain the latest mouse position
When a drag is started or the mouse is moved, identify the subpart it corresponds to and then change the cursorWhenDragged or cursorWhenMoved property.

So I'm not sure for `cursorWhenMouseOver`, and for `cursorWhenDragged` I don't think it can work. Once dragging is on the cursor could go anywhere.

abaga129 commented:
Ah okay. Perhaps I'm misunderstanding some fundamentals or over complicating it.  It may be better to return it to how I initially had it with private properties of `_cursorWhenMouseOver` and `_cursorWhenDragged`. We could just update cursor  from `onMouseMove` and `onMouseDrag`

abaga129 commented:
I've reverted most of the previous commit and I've added a special invisible cursor to be used for `MouseCursor.hidden`

Here is a gif of the current state of things in X11.  
![Destructatorizer Gif](https://imgur.com/a/N5lRKS1)
I'm fairly happy with this! I'm moving on to win32 now.
